### DISCUSSION TITLE
How to get user name by id in the `EventListenerProvider`

### DISCUSSION THREAD

**Discussion by missedone** [keycloak/keycloak#29232] [Q&A] [answered]
hi
i'm implementing the `EventListenerProvider` to forward the admin events to SIEM system.

The implementation works well except that i need a user name rather than `userId` in the payload, is there any method or API for me to get the user name by user ID in the EventListenerProvider?

```
"data": {
      "realmId": "5384fc72-1822-438d-96ee-a851b3f93e5f",
      "resourcePath": "events/config",
      "operationType": "UPDATE",
      "time": 1714622352590,
      "id": "ab1b444f-8e69-47ec-98bf-7ccb16fed01a",
      "authDetails": {
        "clientId": "67ec591c-fea2-456e-807f-b94f4620f627",
        "realmId": "29733b7c-f52d-42c1-af46-5cf214cab17f",
        "ipAddress": "100.126.48.101",
        "userId": "5979700a-a5cc-47f2-8816-0a389519a37b"
      },
      "resourceTypeAsString": "REALM",
      "resourceType": "REALM"
    },
```

**Comment by codespearhead** [comment]
What if the user has no username or changes their username? What if the user uses [their email as their username](https://github.com/keycloak/keycloak/issues/12573), then changes the email?

The only attribute that cannot be changed is **their id, which should be used to avoid leaking private information** and facilitating the existence of untrackable logs and bugs.

Either way, if you're doing so for learning purposes, here are two a quick implementations from [ChatGPT-4 and Google's Gemini](https://github.com/keycloak/keycloak/files/15192504/discussion-29232.md). I haven't tested them, but considering the methods they used I believe it should work.

**Comment by missedone** [comment]
the AdminEvent doesn't have `getSession()` method https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/events/admin/AdminEvent.java

**Reply by codespearhead** [reply, marked as answer]
True. Have you tried implementing the [create method](https://github.com/keycloak/keycloak/blob/5e00fe8b10d51b17f0ab60632720e88fe8e94659/server-spi/src/main/java/org/keycloak/provider/ProviderFactory.java#L36) from [EventListenerProviderFactory](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/events/EventListenerProviderFactory.java)?

```java
public class MyProviderFactory implements EventListenerProviderFactory {
    @Override
    public EventListenerProvider create(KeycloakSession session) {
        return new MyProvider(session);
    }
}
```

```java
public class MyProvider implements EventListenerProvider {
    private KeycloakSession session;

    public MyProvider(KeycloakSession session) {
        this.session = session;
    }

    @Override
    public void onEvent(AdminEvent event) { ... }
}
```

**Reply by missedone** [reply]
this is good idea.

**Reply by missedone** [reply]
i just tried and it worked. thanks

**Comment by NagenderPulluri** [comment]
@missedone  Hi, I'm having similar kind of situation, Can you please help how did you resolve this.

I have enabled the keycloak event logs by using ENV variable  "KC_SPI_EVENTS_LISTENER_JBOSS_LOGGING_SUCCESS_LEVEL"  with value "info"
after this I'm getting event logs as below for e.g
2024-05-08 07:24:43,646 INFO  [org.keycloak.events] (executor-thread-14) type=LOGIN, realmId=cdmob, clientId=cdmobui, userId=340GHJf47-12d5-4397-a4gf1-cghj2a247095, ipAddress=1xx.21.63.xxx, auth_method=openid-connect, auth_type=code, response_type=code, redirect_uri=https://abc-xyz.com/auth/callback, consent=no_consent_required, code_id=39u87dcgdefe-678a-4f9b-7843-b3b72d582d5b, username=johnson, response_mode=query, authSessionParentId=39hjkbbefe-424a-4f9b-ghj57-b3b72d582d5b, authSessionTabId=QYHH64D6mBD49k

I would like to add a new custom attribute "appName" to the above logs, as I'm sending these logs to some other external tool which is expecting "appName" along with other attributes like realmId, clientId, userId, ipAddress etc in the above logs.  We are mentioning our application name in client description while creating client in realm. so I would like to pass this client description to the attribute "appName" and that would be to be added to above event logs.

Can you please help me how do I achieve adding custom attribute to the event logs? Thanks in advance.

**Reply by codespearhead** [reply]
You haven't provided any feedback in your original question (#29382).

If the proposed solution there didn't work, please provide a follow-up there instead of commenting same question in other discussions.
